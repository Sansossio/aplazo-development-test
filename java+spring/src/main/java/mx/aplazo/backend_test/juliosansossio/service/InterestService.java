package mx.aplazo.backend_test.juliosansossio.service;

import java.util.List;

import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterest;
import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterestResponse;
import mx.aplazo.backend_test.juliosansossio.domain.Interest;

public interface InterestService {
  public List<Interest> list();

  public Interest save(Interest interest);

  public List<CalculateInterestResponse> calculate(CalculateInterest body);
}
