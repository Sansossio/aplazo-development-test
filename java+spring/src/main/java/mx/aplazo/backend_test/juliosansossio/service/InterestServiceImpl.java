package mx.aplazo.backend_test.juliosansossio.service;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterest;
import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterestResponse;
import mx.aplazo.backend_test.juliosansossio.domain.Interest;
import mx.aplazo.backend_test.juliosansossio.repository.InterestRepository;

@Service
@Validated
public class InterestServiceImpl implements InterestService {
    @Autowired
    private InterestRepository interestRepository;

    public List<Interest> list() {
        return interestRepository.findAll(Sort.by(Sort.Order.desc("id")));
    }

    public Interest save(@NotNull @Valid Interest interest) {
        return interestRepository.save(interest);
    }

    public List<CalculateInterestResponse> calculate(@Valid CalculateInterest body) {
        List<CalculateInterestResponse> response = new ArrayList<>();

        Double amount = body.getAmount();
        Integer terms = body.getTerms();
        Double rate = body.getRate();

        Double totalToPay = amount + (rate / 100) * amount;
        Double totalPerPayment = totalToPay / terms;

        for (int i = 1; i <= terms; i++) {
            CalculateInterestResponse calculateInterestResponse = new CalculateInterestResponse();
            calculateInterestResponse.setPayment_number(i);
            calculateInterestResponse.setAmount(totalPerPayment);
            calculateInterestResponse.setPayment_date(new Date());
            response.add(calculateInterestResponse);
        }

        return response;
    }
}
