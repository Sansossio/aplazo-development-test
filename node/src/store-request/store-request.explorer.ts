import { Injectable, OnModuleInit } from '@nestjs/common';
import { InstanceWrapper } from '@nestjs/core/injector/instance-wrapper';
import { DiscoveryService, MetadataScanner, Reflector } from '@nestjs/core';
import { STORE_REQUEST_KEY } from './decorator/store-request.const';
import { StoreRequestService } from './service/store-request.service';

@Injectable()
export class StoreRequestExplorer implements OnModuleInit {
  constructor(
    private readonly service: StoreRequestService,

    private readonly discoveryService: DiscoveryService,
    private readonly metadataScanner: MetadataScanner,
    private readonly reflector: Reflector,
  ) {}

  private lookupHandlers(instance: Record<string, Function>, key: string) {
    const methodRef = instance[key];
    const config = this.reflector.get(STORE_REQUEST_KEY, methodRef);

    if (!config) {
      return;
    }

    instance[key] = async (...args: any[]) => {
      const response = await methodRef(...args);
      if (args.length === 1) {
        [args] = args;
      }
      void this.service.storeEvent(args, response);
      return response;
    };
    instance[key].bind(instance);
  }

  onModuleInit() {
    const instances: InstanceWrapper[] = [
      ...this.discoveryService.getControllers(),
      ...this.discoveryService.getProviders(),
    ];

    instances
      .filter((wrapper) => wrapper.isDependencyTreeStatic())
      .forEach(({ instance }: InstanceWrapper) => {
        if (!instance || !Object.getPrototypeOf(instance)) {
          return;
        }
        this.metadataScanner.scanFromPrototype(
          instance,
          Object.getPrototypeOf(instance),
          (key: string) => this.lookupHandlers(instance, key),
        );
      });
  }
}
