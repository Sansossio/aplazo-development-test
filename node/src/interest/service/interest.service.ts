import { Injectable } from '@nestjs/common';
import { StoreRequest } from '../../store-request/decorator/store-request.decorator';
import { CalculateInterestRequestDto } from '../dto/calculate.request.interest.dto';
import { CalculateInterestResponseDto } from '../dto/calculate.response.interest.dto';

@Injectable()
export class InterestService {
  @StoreRequest()
  calculateInterest(
    data: CalculateInterestRequestDto,
  ): CalculateInterestResponseDto[] {
    const response: CalculateInterestResponseDto[] = [];
    const totalToPay = data.amount + (data.rate / 100) * data.amount;
    const totalPerPayment = totalToPay / data.terms;

    for (let i = 1; i <= data.terms; i++) {
      response.push({
        payment_number: i,
        amount: totalPerPayment,
        payment_date: new Date(),
      });
    }
    return response;
  }
}
