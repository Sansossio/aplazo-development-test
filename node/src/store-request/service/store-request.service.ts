import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { StoreRequestEntity } from '../../database/models/store-request.entity';

const LIMIT = 10;

@Injectable()
export class StoreRequestService {
  constructor(
    @InjectRepository(StoreRequestEntity)
    private readonly storeRequestRepository: Repository<StoreRequestEntity>,
  ) {}

  async storeEvent(body: Object, response: Object) {
    const storeRequest = this.storeRequestRepository.create({
      request: JSON.stringify(body),
      response: JSON.stringify(response),
    });
    await this.storeRequestRepository.save(storeRequest);
  }

  async getEvents() {
    return this.storeRequestRepository.find({
      take: LIMIT,
      order: {
        id: 'DESC',
      },
    });
  }
}
