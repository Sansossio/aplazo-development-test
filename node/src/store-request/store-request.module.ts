import { Module } from '@nestjs/common';
import { DiscoveryModule } from '@nestjs/core';
import { TypeOrmModule } from '@nestjs/typeorm';
import { StoreRequestEntity } from '../database/models/store-request.entity';
import { StoreRequestExplorer } from './store-request.explorer';
import { StoreRequestService } from './service/store-request.service';
import { StoreRequestController } from './controller/store-request.controller';

@Module({
  imports: [TypeOrmModule.forFeature([StoreRequestEntity]), DiscoveryModule],
  controllers: [StoreRequestController],
  providers: [StoreRequestService, StoreRequestExplorer],
})
export class StoreRequestModule {}
