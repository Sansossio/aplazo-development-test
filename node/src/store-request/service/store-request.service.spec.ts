import { StoreRequestService } from './store-request.service';

describe('Store Request Service', () => {
  let service: StoreRequestService;
  const repository = {
    create: jest.fn(),
    save: jest.fn(),
    find: jest.fn(),
  };

  beforeEach(() => {
    jest.resetAllMocks();
    service = new StoreRequestService(repository as any);
  });

  describe('Get events', () => {
    it('should return the events', async () => {
      const events = [
        {
          id: 1,
          request: '{}',
          response: '{}',
        },
        {
          id: 2,
          request: '{}',
          response: '{}',
        },
      ];
      repository.find.mockImplementation(async () => events);

      expect(await service.getEvents()).toMatchSnapshot();
    });

    it('should match filters', async () => {
      let response = null;
      repository.find.mockImplementation(async (filters) => {
        response = filters;
      });
      await service.getEvents();
      expect(response).not.toEqual(null);
      expect(response).toMatchSnapshot();
    });
  });

  describe('Store event', () => {
    it('should call create and save methods 1 time', async () => {
      await service.storeEvent({}, {});
      expect(repository.save).toHaveBeenCalledTimes(1);
      expect(repository.create).toHaveBeenCalledTimes(1);
    });

    it('should returns undefined', async () => {
      expect(await service.storeEvent({}, {})).toBeUndefined();
    });
  });
});
