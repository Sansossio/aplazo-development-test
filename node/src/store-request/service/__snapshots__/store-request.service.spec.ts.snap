// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Request Service Get events should match filters 1`] = `
Object {
  "order": Object {
    "id": "DESC",
  },
  "take": 10,
}
`;

exports[`Store Request Service Get events should return the events 1`] = `
Array [
  Object {
    "id": 1,
    "request": "{}",
    "response": "{}",
  },
  Object {
    "id": 2,
    "request": "{}",
    "response": "{}",
  },
]
`;
